### ISSUE TITLE
Enhancement: Support For Live Preview / Preview Links

### ISSUE THREAD

**Issue by edolyne** [markhuot/craftql#31]
It would be great if there was a way to enable graphql to work via live preview and preview links.

For preview links, potentially leveraging the token as an argument to pass which can then fetch the draft or return an error if the token has expired.

For example:

```graphql
{
  entry(slug: "test", token: "gO9vhGh4WgicL-bjUr78VKxT7MMw2zqW") {
    title
    id
  }
}
```

For live preview, I'm not quite sure about a proposed solution as I'm not sure if or how craft stores the live preview.

**Comment by markhuot** [comment]
I’ll need to look into this. The last I checked, at least in Craft 2, preview data was not exposed through the API. It was only sent as POST data directly to the entry template, which we don’t have access to query directly.

**Comment by markhuot** [comment]
After looking in to this more I don't think it's possible for CraftQL, alone, to manage live preview URLs. The way Craft POSTs preview data directly to a template makes it impossible for CraftQL to intercept. You, could, however use the `drafts` field on and `EntryConnection` to pull Draft's out of the database to approximate this experience.

**Comment by timkelty** [comment]
I looks like Live Preview will become tokenized in the future, which may help: https://github.com/craftcms/cms/issues/1521

**Comment by timkelty** [comment]
> I'll browse through the https://github.com/craftcms/cms issue list and see if moving previews away from POST data is on the horizon and if not I'll open an issue to ask about it.

See @narration-sd's reply to my comment: https://github.com/craftcms/cms/issues/1521#issuecomment-409895388

**Comment by oddnavy** [comment]
Is it now possible to access live preview content using the "Tokenized Live Preview authorization" added in craftcms/cms#1521?

Testing locally it seems like if I add a valid `X-Craft-Token` header, the graphql api attempts to return the twig template for the entry rather than the JSON. If I remove the `X-Craft-Token` header the request works prefectly.

**Comment by markhuot** [comment]
I’m looking in to this more. I don’t think the `X-Craft-Token` gets us preview content, that content is still hidden away in a POST request. But I’m working on an update to CraftQL that’ll store that POST data and make it available to GraphQL when the header token is present.

I’m going to re-open this ticket because I have an idea on how to implement this.

**Comment by narration-sd** [comment]
@oddnavy @markhuot  Actually, that's just right, Mark. There is no data present  for CraftQL to see during live updates.

I've got a full solution about ready to go out the door. It takes much more than you may imagine to do this right -- essential safety, efficiency, and flexibility as well as correctness, and then get it hooked in a useful fashion to SPA or  embedded-component sites. 

And the data availability side is far from the only issue in providing something others can readily work with in practice.

What I've got has properly  had a long gestation. It's called Live Vue, is a Craft plugin which will be priced like yours, and will go out initially for Gatsby, which  has its own issues, and drew me into one more reconstitution of the architecture, 

I'm doing it this way for all platforms now because I like the design result Gatsby teased me into. Much for the developer using it is simplified, achieving that goal.

This will finally close off a large area of issues which I had solved and refined to a limit, but still concerned I feel rather accurately about how much trouble it would be in terms of handholding, to field the original. 

Which has long been working solidly on Vue, where it was first worked up, for  many months. The Gatsby effort gets me also to quite solid working with React as well. I'm proceeding through a limited beta as much for community 'chiefs' confidence as much as to see that readiness is there.  Not long now.

**Comment by narration-sd** [comment]
...and by the way, I should say thanks again, for CraftQL. This thing started out on Element API, and then had some 'interesting' and potential abilities, but once I saw how clear it was to put together many things persons ordinarily want to do, I was sold on your path. It'll be a boon to many, and has already been here.

**Comment by monachilada** [comment]
The tokenized preview feature has now been added to the Craft 3.2 alphas! :) See commit: https://github.com/craftcms/cms/commit/197ce312dd641dcad5838fcb442dfef27da538c1

From @brandonkelly's description:

> Preview requests will have a token query string parameter, and as long as you pass that along to your Element API (or other API) request, the draft content will be returned instead of the main entry content.

This works great in terms of what gets passed to the Live Preview iframe, but I'm struggling to figure out in which way that token needs to be passed through my CraftQL queries or requests to get it to work, or even whether changes need to be made to the plugin itself to enable it. Can you chime in @markhuot?

I'd be happy to help out if I can, just let me know.

**Comment by brandonkelly** [comment]
@monachilada Add something like this to your JS:

```javascript
// Get the preview token from the URL
let m = document.location.href.match(/\btoken=([^&]+)/);
let token = m ? m[1] : '';

// Then forward that on whenever you are sending a CraftQL API request
let url = '...';
if (token) {
    url += `?token=${token}`;
}
// ...
```

If you’ve set a custom [tokenParam](https://docs.craftcms.com/v3/config/config-settings.html#tokenparam) config setting, use that instead of `token` here.

CraftQL uses element queries internally to fetch its content, so in theory this will just work, and you’ll start seeing draft content sent back when a token is present.

**Comment by SnarkieDesign** [comment]
@brandonkelly 🤔 I'm getting CORS issues when forwarding the token to my SPA. Happening on both 3.2 beta and alpha
```
Access to fetch at 'https://localhost/api?token=MtitB8j722_Qwo45bYOAdDXKQmfM-hJT' from origin 'https://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by brandonkelly** [comment]
@SnarkieDesign I’ve just disabled CSRF validation for preview requests, for the next 3.2 Beta release.

To get the fix early, change your `craftcms/cms` requirement in composer.json to:

```json
"require": {
  "craftcms/cms": "3.2.x-dev#54831ac139c954f2499fa81b977fed70803f8792 as 3.2.0-beta.2",
  "...": "..."
}
```

Then run `composer update`. And then you should be able to re-enable your `enableCsrfProtection` config setting.

**Comment by codyjames** [comment]
Has anybody gotten CraftQL to work with the new Headless Preview and draft set-up?

**Comment by narration-sd** [comment]
https://github.com/craftcms/cms/issues/4584#issuecomment-512569473

**Comment by codyjames** [comment]
@brandonkelly Is the goal here that we won't have to write specific _draft_ queries, but rather we'd just have to include the token in the request and it'd return the associated draft?

**Comment by brandonkelly** [comment]
@codyjames Yep that’s the goal. It’s not perfect yet for unsaved/disabled entries (see craftcms/cms#4581), but we’re getting there.

**Comment by codyjames** [comment]
@brandonkelly Okay great, that'll be so great. Yeah, I was running into the unsaved/disabled issue. It'd be _fantastic_ if the token made it so the associated entry was considered "enabled" when the token was present.

**Comment by brandonkelly** [comment]
Yeah, that’s the general idea of how we could solve this. For unsaved drafts it’s a little more complicated because we also have to let it ignore the `draft` status for that one element.

**Comment by brandonkelly** [comment]
@codyjames Craft 3.2.5 should fix this. Details and info on how you can test here: https://github.com/craftcms/cms/issues/4581#issuecomment-513026880

**Comment by codyjames** [comment]
@brandonkelly Okay, this is working for me. I love it. We're using it for a Next.js app and it is working like a charm.

**Comment by brandonkelly** [comment]
Awesome!

**Comment by brandonkelly** [comment]
By the way, the official 3.2.5 release is out now, which fixes a bug introduced by my original commit referenced above. So make sure you’re running that.

**Comment by Jones-S** [comment]
I see there has been some progress on this issue. But what I don't really know:
1. is it possible to get the newest draft via GraphQL by adding this new token from craft
2. If so, how does a request look like. I can't my head get around how I should send that token. Also because my apollo client already sends the Bearer token to get access to the graphql service.

Thanks in advance.
cheers

**Comment by lukehmu** [comment]
> 
> 
> I see there has been some progress on this issue. But what I don't really know:
> 
>     1. is it possible to get the newest draft via GraphQL by adding this new token from craft
> 
>     2. If so, how does a request look like. I can't my head get around how I should send that token. Also because my apollo client already sends the Bearer token to get access to the graphql service.
> 
> 
> Thanks in advance.
> cheers

Hi @Jones-S - I'm working on a Vue project where I'm using the CraftQL plugin. I have _just_ got preview working correctly.

As @brandonkelly said [here](https://github.com/markhuot/craftql/issues/31#issuecomment-499062235) you need to apend the craft token as a query parameter onto the request URL.

Excuse the syntax, I'm using Vue Router to get my query params & I've rewritten it a little to be more readable - it's a ternary in my code:
```JavaScript
// grab the API URL from my env file
let postURL = process.env.API_URL

// if my app is requested with the x-craft-preview query param
if (router.currentRoute.query['x-craft-preview']) {
  // update the URL I am going to use in my POST to include the token from the query params.
  const postURL = `${process.env.API_URL}?token=${router.currentRoute.query.token}`
}
```


Then I send it off via Axios in the regular way (I tried using Axios params, but it didn't work, hence changing the URL)
```JavaScript
await axios.post(postURL, {
      query: myGraphQLQuery,
    }).then // other stuff
```

[Lnk to where I do this in my code, hope this helps!](https://github.com/lukehmu/vue-webpack/blob/master/src/js/store/index.js)

**Comment by Jones-S** [comment]
@lukehmu 
Thanks for your hint.
I myself got some kind of a working thing.
I checked for the x-craft-preview param as well and would then send another graphql query.
Where I would normally fetch a specific post I woudl this time go for a draft:

```
gql`query getDraft($slug: String!) {
      entriesConnection(slug: $slug) {
        edges {
          node {
            id
            title
          }
          drafts {
            edges {
              draftInfo {
                draftId
                name
              }
              node {
                id
                title
                ... on News {
                  body {
                    totalPages
                    content
                  }
                }
              }
            }
          }
        }
      }
    }`;
```
Here I fetch drafts and then I would dive into the node (that's where the actual data lies).
After getting that data I would just take the latest of the drafts.

If your option works, this would be great of course and much nicer.
So if I understand correctly:
By sending a graphql request via axios, while passing on the token, I could use the normal query for my page and I would automatically get the latest draft?...

**Comment by lukehmu** [comment]
@Jones-S I started to go down this route, too.

However, passing the token as a query string in the request URL works perfectly for me.

You can test this easily by previewing your entry, inspecting the preview, looking at the request URL and getting the token. Then hardcode the token in your app and you'll see the draft version! Then refactor to do it dynamically as per my example above.

Cheers.

**Comment by Jones-S** [comment]
@lukehmu 
I see what you are doing and I am just trying to replicate in my setup.
My setup though makes use of apollo (nuxt apollo module). Would you have any hints on how to approach that?

Because normally if someone enters an url it would fetch the page via apollo normally.
Now I added some `middleware` that checks for the crafts preview token.
If present I want to get the draft data by adding the craft token, but I don't know how I would do that with an apollo call...

**Comment by ca-miked** [comment]
@Jones-S I am running into the same issue.

**Comment by Jones-S** [comment]
@ca-miked You might want to check out craft cms 3.3 with graphl in its core. It will change a few things. You can set a headlessMode and in 3.4 there will be preview Targets: https://github.com/craftcms/cms/issues/4520

I hope it will solve the thing about the slugs of newly created posts. I am currently trying out the new setup.

**Comment by Jones-S** [comment]
@ca-miked indeed it is fixed. With Craft 3.4.x-dev (maybe even before) I can create a new entry, and as soon as I have set a title (and therefore a slug) it will open the preview with the correct slug, even if I haved not saved it before. 👍🏼
